doc/DifferentiableManifoldRecipe.md

```diff
@@ -0,0 +1,128 @@
+# Differentiable Manifold Recipe
+
+This recipe describes how to express a differentiable manifold in Swift. The
+main payoffs of using this recipe are:
+
+ * You can choose arbitrary global and local coordinate systems for your
+   manifolds.
+ * Swift's Automatic Differentiation works with the coordinate system that you
+   specify. In particular, the tangent vectors that flow through the
+   automatically-generated derivatives are exactly the tangent vectors
+   corresponding to the local coordinate system you specify.
+
+## Steps
+
+### 1. Choose a global coordinate system.
+
+Create a `struct` whose values represent points on the manifold. Every point on
+the manifold must be representable by at least one `struct` value. Every
+`struct` value must represent at most one point on the manifold.
+
+For example, both of these `struct`s are acceptable global coordinate systems
+for `SO(2)`, the manifold of 2D rotations:
+
+```swift
+struct SinCosGlobalCoordinate {
+  /// The cosine of the angle of the rotation.
+  var c: Float
+
+  /// The sine of the angle of the rotation.
+  var s: Float
+}
+
+struct AngleGlobalCoordinate {
+  /// The angle of the rotation.
+  var theta: Float
+}
+```
+
+### 2. Choose the local coordinate type.
+
+Choose an `n`-dimensional vector type to represent to represent the manifold's
+local coordinates.
+
+For example, `Float` and `Vector1` are both acceptable types for `SO(2)` local
+coordinates.
+
+### 3. Conform to the ManifoldCoordinate protocol.
+
+The coordinate `struct` you chose in step 1 should conform to the
+`ManifoldCoordinate` protocol, which has these requirements:
+
+```swift
+/// A coordinate in a manifold's global coordinate system.
+public protocol ManifoldCoordinate: Differentiable {
+  /// The global coordinate corresponding to `local` in the chart centered around `self`.
+  @differentiable(wrt: local)
+  func global(_ local: LocalCoordinate) -> Self
```

Comment:
Hi Frank!

Question about exponential map: the first version you shared (`Vector3 -> Rot3`) is an unary function, but I understand exponential map to be a binary function: `(T, T.TangentVector) -> T`.

Could you please clarify what the first version means?

Reply:
I'm not completely sure I understand your comment. But I'm very open to finding a better name for these things.

> I got confused between what was mutating and what is functional.

None of the requirements in this file are mutating. (There is the `move(along:)` requirement in `Differentiable` which is a mutating version of exp.)

> Rot3::Expmap: Vector3 -> Rot3 (functional): v -> exp(v)

By `Rot3 (functional)`, do you mean that you're thinking of the `Rot3` as a function by the group action on itself?

If I've interpreted that correctly, we can't have an `ExpMap: Vector3 -> Rot3 (functional)` for a general `Manifold` because the general `Manifold` doesn't have a group operation.

We could of course have an `ExpMap`-like function on `LieGroup` (if we eventually define a `LieGroup`).

> Rot3::expmap: Rot3 * Vector3 -> Rot3 (still functional, but also composes): R, v -> R*exp(v)

This is the one that corresponds most closely to the current `ManifoldCoordinate.global` in this PR. That's because `ManifoldCoordinate.global` is a method that takes a `self` parameter (the base point for the exp map). e.g. it's signature for `Rot3Coordinate` is `Rot3Coordinate.global: (Rot3Coordinate) -> (Vector3) -> Rot3Coordinate`.

Using "exponential map" in the names here might not be a good idea because this is a general manifold, not a lie group.

Here's another name idea:
* `global` -> `retraction`
* `local` -> `inverseRetraction`

Reply:
`Expmap` is a unary function which is around the origin. It is really closest to the mathematical concept: all matrix groups define an exponential map with is really the matrix exponential, and it is a map from Lie algebra to Lie group. Unary.

 `expmap`, lowercase, is really the "right-trivialization" (if I'm not messing up this definition - always a bit mysterious): we compose a rotation on the right with the exponential map.

Reply:
@marcrasi you are right: manifolds don't necessarily have a group operation - let me regroup (no pun intended!)

Reply:
So, in GTSAM, we call `global` just `retract`, which matches Absil book.

Reply:
but I never liked `localCoordinates` all that much. I'm not in love with `inverseRetract` either, though.

Reply:
To amend my comment to allow for manifolds: *if* T is a Lie group, both Expmap and expmap exist, we implement `retract` with `expmap`. If not, we do something manifold-specific.

Reply:
`retract` (and `inverseRetract`?) would work for me!

It would be slightly more "swifty" to name them `retracted` and `inverseRetracted` because of the guideline "When the operation is naturally described by a verb, use the verb’s imperative for the mutating method and apply the “ed” or “ing” suffix to name its nonmutating counterpart." (https://swift.org/documentation/api-design-guidelines/).

But I wouldn't object to calling it `retract` for consistency with existing terminology.

Reply:
Other random name ideas: `unretract`, `localized`.

> To amend my comment to allow for manifolds: if T is a Lie group, both Expmap and expmap exist, we implement retract with expmap. If not, we do something manifold-specific.

Automatically implementing `retract` with `expmap` for Lie groups could be done with protocol extensions in Swift like this:

```
protocol LieGroupCoordinates: ManifoldCoordinates {
  func expmap(_ v: LocalCoordinates) -> Self
}

extension LieGroupCoordinates {
  func retract(_ v: LocalCoordinates) -> Self {
    expmap(v)
  }
}
```